### PR TITLE
2.0 Event API

### PR DESCRIPTION
**Pull request by eile**
Review - do not merge yet

### CODE REVIEW

File: eq/channel.h [hunk 1]
```diff
@@ -350,11 +349,15 @@ class Channel : public fabric::Channel< Window, Channel >
      * transform the event into a config event to be send to the application
      * using Config::sendEvent().
      *
+     * @param type the pointer event type.
      * @param event the received event.
      * @return true when the event was handled, false if not.
      * @version 1.0
      */
-    EQ_API virtual bool processEvent( const Event& event );
+    EQ_API virtual bool processEvent( EventType type, SizeEvent& event );
```
Comment (tribal-tec) [inline eq/channel.h]: event object cannot be const-ref?

File: eq/config.cpp [hunk 2]
```diff
@@ -641,130 +593,196 @@ EventICommand Config::getNextEvent( const uint32_t timeout ) const
...
-            if( event.pointerButtonPress.buttons ==
-                ( PTR_BUTTON1 | PTR_BUTTON2 | PTR_BUTTON3 ))
-            {
-                _impl->running = false;
-                return true;
-            }
-            break;
+bool Config::handleEvent( const EventType type, const AxisEvent& )
+{
+    switch( type )
```
Comment (tribal-tec) [inline eq/config.cpp]: only 'return false'?

File: eq/config.cpp [hunk 3]
```diff
@@ -641,130 +593,196 @@ EventICommand Config::getNextEvent( const uint32_t timeout ) const
...
+    }
+}
 
-        case Event::STATISTIC:
-            LBLOG( LOG_STATS ) << event << std::endl;
-            addStatistic( event.serial, event.statistic );
-            break;
+bool Config::handleEvent( const EventType type, const ButtonEvent& )
+{
+    switch( type )
```
Comment (tribal-tec) [inline eq/config.cpp]: dito?

File: eq/config.h [hunk 4]
```diff
@@ -429,13 +376,19 @@ class Config : public fabric::Config< Server, Config, Observer, Layout, Canvas,
...
 
     /**
-     * Handle one config event. Thread safe.
+     * Handle the given event. Thread safe.
      *
      * @param command the event command.
      * @return true if the event requires a redraw, false if not.
      * @version 1.5.1
      */
     EQ_API virtual bool handleEvent( EventICommand command );
```
Comment (tribal-tec) [inline eq/config.h]: Is the generic EventICommand handleEvent something an application would need/want to use? Or the new funcs are enough?

File: eq/deflect/eventHandler.cpp [hunk 5]
```diff
@@ -178,80 +169,97 @@ void EventHandler::_processEvents( const Proxy* proxy )
...
+        {
+            KeyEvent event;
+            event.originator = channel.getID();
+            event.serial = channel.getSerial();
+            const EventType type =
+                deflectEvent.type == ::deflect::Event::EVT_KEY_PRESS ?
+                                         EVENT_KEY_PRESS : EVENT_KEY_RELEASE;
+            event.key = _getKey(  deflectEvent.key );
+            channel.processEvent( type, event );
+            return;
```
Comment (tribal-tec) [inline eq/deflect/eventHandler.cpp]: even though that's what the old code was doing with 'break', I'm wondering it should be 'continue' to handle all pending events?

File: eq/fabric/event.h [hunk 6]
```diff
@@ -20,221 +20,41 @@
...
-        MagellanEvent magellan;           //!< SpaceMouse data
-
-        UserEvent     user;               //!< User-defined event data
-    };
+    Event() : serial(0), time(0) {}
 
-    /** The last rendering context for the pointer position. */
-    RenderContext context;
+    uint32_t serial; //!< unique originator serial number
+    int64_t time; //!< The config time when the event was created
```
Comment (tribal-tec) [inline eq/fabric/event.h]: signed needed?

File: eq/node.cpp [hunk 7]
```diff
@@ -467,13 +469,24 @@ void Node::frameTasksFinish( const uint128_t&, const uint32_t frameNumber )
...
-    return getConfig()->sendError( Event::NODE_ERROR, Error( error, getID( )));
+    return getConfig()->sendError( EVENT_NODE_ERROR, Error( error, getID( )));
 }
 
-bool Node::processEvent( const Event& event )
+bool Node::processEvent( const EventType type, AxisEvent& event )
 {
-    ConfigEvent configEvent( event );
-    getConfig()->sendEvent( configEvent );
+    getConfig()->sendEvent( type ) << event;
```
Comment (tribal-tec) [inline eq/node.cpp]: feels having sendEvent() part of node etc. to omit the superfluous getConfig()...

File: eq/qt/event.cpp [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    case Qt::Key_F22:       return KC_F22;
+    case Qt::Key_F23:       return KC_F23;
+    case Qt::Key_F24:       return KC_F24;
+    case Qt::Key_Shift:     return KC_SHIFT_L;
+    case Qt::Key_Control:   return KC_CONTROL_L;
+    case Qt::Key_Alt:       return KC_ALT_L;
+    case Qt::Key_AltGr:     return KC_ALT_R;
+    case Qt::Key_unknown:   return KC_VOID;
+    default:
+        if( keyEvent.text().isEmpty( ))
```
Comment (tribal-tec) [inline eq/qt/event.cpp]: check master. This code misses the magic '+32'

File: eq/server/config.cpp [hunk 9]
```diff
@@ -846,12 +846,10 @@ bool Config::exit()
...
-    // TODO: is this needed? sender of CMD_CONFIG_EXIT is the appNode itself
-    // which sets the running state to false anyway. Besides, this event is
-    // not handled by the appNode because it is already in exiting procedure
-    // and does not call handleEvents anymore
-    // eile: May be needed for reliability?
-    send( findApplicationNetNode(), fabric::CMD_CONFIG_EVENT ) << Event::EXIT;
+    // send exit event to app, needed if this is called from init()
+    EventOCommand cmd( send( findApplicationNetNode(),
+                             fabric::CMD_CONFIG_EVENT ));
+    cmd << EVENT_EXIT;
```
Comment (tribal-tec) [inline eq/server/config.cpp]: send is only finished after cmd is out of scope, no?

File: eq/systemWindow.h [hunk 10]
```diff
@@ -152,8 +152,23 @@ class SystemWindow
      */
     EQ_API EventOCommand sendError( const uint32_t error );
 
-    /** Process an event. @version 1.0 */
-    EQ_API virtual bool processEvent( const Event& event );
+    /** Process a stateless event. @return true if the event was handled. */
+    EQ_API bool processEvent( EventType type );
```
Comment (tribal-tec) [inline eq/systemWindow.h]: not virtual for GLWindow?

**Comment by tribal-tec** [comment]
I'm finished with it literally. There seemed to be an amendment in the meantime, but I don't know what you changed there. You either tell me or make new commits to isolate those changes.

File: eq/channel.h (hunk 1 above)
Comment (eile) [inline eq/channel.h]: Updated doxygen to:
```
     * The task of this method is to update the channel and event as necessary,
     * and send it to the application using Config::sendEvent().
```

Aka there are use cases where the event is modified in place during processing, e.g., to set the originator ID to the channel id.

File: eq/config.cpp (hunk 2 above)
Comment (eile) [inline eq/config.cpp]: yes.

File: eq/config.h (hunk 4 above)
Comment (eile) [inline eq/config.h]: Yes, for EVENT_USER or EVENT_OBSERVER_MOTION, which do not have a more specialized handleEvent( type, struct ).

File: eq/server/config.cpp (hunk 9 above)
Comment (eile) [inline eq/server/config.cpp]: yes - bit sure what your point is?

**Comment by eile** [comment]
No amendment yet, just rebased onto master.

File: eq/server/config.cpp (hunk 9 above)
Comment (tribal-tec) [inline eq/server/config.cpp]: Then the new code is different than before. Not sure if this has any effect.

**Comment by tribal-tec** [comment]
Not sure it was mentioned why the Qt events have to be clone when reposted (better put it for later reference).
+1 otherwise, unless something is still missing/pending.

**Comment by tribal-tec** [comment]
PR contains two of my commits. Rebase again?

**Comment by tribal-tec** [comment]
+1 otherwise

**Comment by eile** [comment]
Retest this please.

**Comment by tribal-tec** [comment]
+1 again

**Pull request merged**